### ISSUE TITLE
ShaderManager

### ISSUE THREAD

**Issue by dlyr** [STORM-IRIT/Radium-Engine#555]
Quick issue tu have a reminder ... ShaderManager miss a remove method, or at least a way to change the configuration of an already added shader.
 Also if a shader "name"is added with multiple configuations, only the first one is accessible with name access.
This issue is here to first list the different use cases we have in mind and then to discuss a possible evolution.

**Comment by MathiasPaulin** [comment]
Things to keep in mind : 
 - Shader manager, in its actual implementation, is badly named and consists on a add-only ProgramShader cache (it do not store shaders but programs : set of linked shaders).
 - The key to access a ProgramShader through the manager is its configuration. Accessing by name first search for a configuration with this name.
 - If several components ask for a ProgramShader with the same configuration, they will all share the same. So, if we modify the configuration, all users of this configuration must be notified so that they can either use the modified configuration or remains with the original one. Observer/observable pattern will help, each users being an observer on the shader configuration. 
 - Doing this will result in name clash as several configurations will have the same name.
 - Shader configurations are always passed and returned by value. There is no reference given on any configuration. So, no one can change an existing configuration, each modification will result in a new ProgramShader with a different configuration (and a name clashing). 
 - Some component, for efficiency reasons, store also the association between a configuration and a shader program. e.g. RenderTechnique do that to access configuration and program in constant time instead of in log time wrt the number of shaders in the ShaderManager (quite high in some application scenario due to add-only policy of the manager)
 - For now, either in RadiumLibs or RadiumApps, accessing shader by name is just do for Engine internal shaders. We can remove name clash easily by making the corresponding method private and declaring Engine friend of the manager.

**Comment by dlyr** [comment]
> For now, either in RadiumLibs or RadiumApps, accessing shader by name is just do for Engine internal shaders. We can remove name clash easily by making the corresponding method private and declaring Engine friend of the manager.

I will agrue against this point: why an "official" radium renderer will have more privilege than a user one ?

**Comment by MathiasPaulin** [comment]
I do not speak about "official renderer" but about the Engine, the GuiBase and the abstract class `Ra::Engine::Renderer` (recall that any renderer have to derived from this class) that uses internal shaders to display textures, render the gui objects, ... 
But if we agree to manage name conflict between internally defined shaders and user defined shaders, then, we must implement something that is robust and not to hard to use.

More generally, the Engine relies in some internal services that can be customized and managed by any renderer. These services must be robust so that one can integrate any renderer into the system without having to (re) develop basic tasks.
These basics services should not be based on external shader files but on inline shaders in the source code. There is no reason to allow the user to modify the .glsl file of the LinearRGB to sRGB conversion before displaying an image.
The Radium libs might proposes services general enough to be used and extended by client applications/libs/plugins but also proposes some advanced services to client that just need them to develop another ones (e.g. if I develop an animation system, I just want to use a renderer that generates nice picture but I don't care on how it works).
And I think that instead of trying to divert the forward renderer from its initial operating way, users have to implement their own but I agree that for now, this is not clearly documented.  But I can give you proof that one can develop a renderer that work in a very different fashion than the forward renderer and that could nevertheless use all the internal Engine services.

**Comment by dlyr** [comment]
So you are voting for *make access by name private and declaring Engine friend of the manager* ?

**Comment by MathiasPaulin** [comment]
Not really. I'm asking _what is the good key to access shaderPrograms: name or configuration?_. _Do we really have to search shaders by name ?_  . Configuration includes name and add other distinctive informations. Configuration collision is far less probable than name collision. So, using configurations as keys will prevent collisions. We can also hash the configuration to make more efficient keys exchange and storage between parts of the engine/application/libs/plugins/...
